src/ORM/Association/BelongsToMany.php

```diff
@@ -343,30 +347,61 @@ public function attachTo(Query $query, array $options = [])
 
         // Attach the junction table as well we need it to populate _joinData.
         $assoc = $this->_targetTable->association($junction->alias());
-        $query->removeJoin($assoc->name());
-        $options = array_intersect_key($options, ['joinType' => 1, 'fields' => 1]);
-        $options += [
+        $newOptions = array_intersect_key($options, ['joinType' => 1, 'fields' => 1]);
+        $newOptions += [
             'conditions' => $cond,
             'includeFields' => $includeFields,
-            'foreignKey' => $this->targetForeignKey(),
+            'foreignKey' => false,
         ];
-        $assoc->attachTo($query, $options);
+        $assoc->attachTo($query, $newOptions);
         $query->eagerLoader()->addToJoinsMap($junction->alias(), $assoc, true);
+
+        parent::attachTo($query, $options);
+
+        $foreignKey = $this->targetForeignKey();
+        $thisJoin = $query->clause('join')[$this->name()];
+        $thisJoin['conditions']->add($assoc->_joinCondition(['foreignKey' => $foreignKey]));
```

Comment:
Well, what solves problems is changing the joins order. This line is so that each of the joins share parts of the conditions. Before, the conditions where always put in the last join